roadmap/implementers-guide/src/node/dispute/votesdb.md

```diff
@@ -0,0 +1,139 @@
+# Votes DB
+
+Responsible module for tracking all kind of dispute votes by all validators
+for a fixed number of sessions.
+
+
+## Remarks
+
+Votes must be persisted. Disputes might happen long after
+there was a dispute, as such just keeping things in memory
+is thus insufficient, as every blip (OOM, bug, maintanance)
+could cause the node to forget the state.
+
+`UnsignedTransaction`s are OK to be used, since the inner
+element `CommittedCanddidateReceipt` is verifiable.
+
+## ToDo
+
+* Missing Definition of Governance Mode.
+* Is a `PoV` enough or should the whole `CommittedCandidateReceipt` be included?
+* Currently [`fn mark_bad()`](https://github.com/paritytech/substrate/pull/6301/files#diff-8faeb5c685a8fdff428c5ec6d9102fd59e127ff69762d43045cd38e586db5559R60-R64) does not persist data.
+
+## IO
+
+Inputs:
+
+* `VotesDbMessage::`
+
+Outputs:
+
+* `DisputeConclusionMessage::Detection`
+* `DisputeConclusionMessage::Resolution`
+* `ChainApiMessage::Blacklist`
+
+## Messages
+
+```rust
+enum VotesDbMessage {
+    /// Allow querying all `Hash`es queried by a particular validator.
+    QueryValidatorVotes {
```

Comment:
What is the use-case of all validator votes for a particular session?

Reply:
It's all votes by a particular validator index. But to disambiguate the validator, we need to know which session index. Which again is necessary since we might have a session change during the dispute.

Reply:
Why don't we just implicitly use the validator set from the session the candidate belongs to so that disputes never have to deal with more than one validator set?

Reply:
Last time I talked to @burdges (iirc) we talked about the last two validator sets being able to dispute, I can't quite recall the reasoning for it though.

Reply:
Eh, it's way easier if we restrict the validator set to the one that was active when that candidate was collated.

Reply:
A priori, I'd favor the validator set when collated/backed, or even when included, over whatever teh current validator set is.  If code turns out simpler some other way then we can always do that.  

There is a risk that a really old validator set may disappear after a while, but not sure this is really a problem.  If so, maybe some manual motion can do it.

Reply:
> A priori, I'd favor the validator set when collated/backed, or even when included, over whatever teh current validator set is. If code turns out simpler some other way then we can always do that.

We should clarify this and nail the definition of `current` in this context.

> There is a risk that a really old validator set may disappear after a while, but not sure this is really a problem. If so, maybe some manual motion can do it.

But if a validator set fades away, it's also not slashable anymore since its staked funds are released as well.

Reply:
The validator set when backed is the one we'll use. "current" is "current" according to the state of whatever relay parent we are executing (runtime) in or inspecting (node-side). The `SessionInfo` module is designed to make this easier.